carousel/core/calculations.py

```diff
@@ -176,31 +83,41 @@ class Calc(object):
     __metaclass__ = CalcBase
 
     def __init__(self):
-        # XXX: Hack to get PR#68 done
-        if not self.parameters:
-            self.parameters = {
-                'always_calc': getattr(self, 'always_calc', False),
-                'frequency': getattr(self, 'frequency', [1, '']),
-                'dependencies': getattr(self, 'dependencies', []),
-                'static': getattr(self, 'static', []),
-                'dynamic': getattr(self, 'dynamic', [])
-            }
+        meta = getattr(self, CalcBase._meta_attr)
+        parameters = getattr(self, CalcBase._param_attr)
         #: ``True`` if always calculated (day and night)
-        self.always_calc = self.parameters.get('always_calc', False)
-        freq = self.parameters.get('frequency', [1, ''])
+        self.always_calc = dict.fromkeys(
+            parameters, getattr(meta, 'always_calc', False)
+        )
+        freq = getattr(meta, 'frequency', [1, ''])
         #: frequency calculation is calculated in intervals or units of time
-        self.frequency = freq[0] * UREG[str(freq[1])]
-        #: list of dependencies
-        self.dependencies = self.parameters.get('dependencies', [])
-        #: list of static calculations
-        self.static = self.parameters.get('static', [])
-        #: list of dynamic calculations
-        self.dynamic = self.parameters.get('dynamic', [])
+        self.frequency = dict.fromkeys(parameters, freq[0] * UREG[str(freq[1])])
+        #: dependencies
+        self.dependencies = dict.fromkeys(
+            parameters, getattr(meta, 'dependencies', [])
+        )
+        #: name of :class:`Calc` superclass
+        self.calc_source = dict.fromkeys(parameters, self.__class__.__name__)
+        #: calculator
+        self.calculator = dict.fromkeys(
+            parameters, getattr(meta, 'calculator', 'static')
+        )
+        #: calcs
+        self.calcs = {}
+        for k, v in parameters.iteritems():
+            self.calcs[k] = {
+                key: v[key] for key in ('formula', 'args', 'returns')
```

Comment:
Good idea, I like it - I'll make an issue ticket to implement a separation of mandatory vs optional parameter/source attributes or meta-options, and how to handle missing mandatory args in the next PR. Thanks!

All of these variables can be set by each parameter inside `CalcParameter`. pls see [L111-L120 in `carousel.core.calculations.Calc`](https://github.com/SunPower/Carousel/blob/953a7d8725e30f39e8ed3ed002a1c657eac4d6f9/carousel/core/calculations.py#L111-L120) of the last commit: 953a7d8725e30f39e8ed3ed002a1c657eac4d6f9

Carousel parameters and their attributes should only be set in `Meta` or a subclass of `Parameters`. Any other class attributes will be treated as personal used by the developer, and ignored by Carousel.